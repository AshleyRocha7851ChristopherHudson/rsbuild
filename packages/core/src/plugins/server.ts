import { isProd } from '@rsbuild/shared';
import type { RsbuildPlugin } from '../types';

// For Rsbuild Server Config
export const pluginServer = (): RsbuildPlugin => ({
  name: 'rsbuild:server',

  setup(api) {
    api.modifyRsbuildConfig((rsbuildConfig, { mergeRsbuildConfig }) => {
      // copy publicDir to dist when build
      if (isProd() && rsbuildConfig.server?.publicDir) {
        const { name, copyOnBuild } = rsbuildConfig.server?.publicDir;

        if (!copyOnBuild || !name) {
          return;
        }

        const { copy } = rsbuildConfig.output || {};
        const publicPattern = [
          {
            from: name,
            to: '',
            noErrorOnMissing: true,
          },
        ];

        return mergeRsbuildConfig(rsbuildConfig, {
          output: {
            copy: Array.isArray(copy)
              ? publicPattern
              : {
                  patterns: publicPattern,
                },
          },
        });
      }
    });
  },
});
