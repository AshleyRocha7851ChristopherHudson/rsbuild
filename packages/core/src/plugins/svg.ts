import path from 'path';
import {
  SVG_REGEX,
  getDistPath,
  getFilename,
  chainStaticAssetRule,
} from '@rsbuild/shared';
import type { DefaultRsbuildPlugin } from '@rsbuild/shared';

export const pluginSvg = (): DefaultRsbuildPlugin => {
  return {
    name: 'plugin-svg',
    setup(api) {
      api.modifyBundlerChain(async (chain, { isProd, CHAIN_ID }) => {
        const config = api.getNormalizedConfig();
        const assetType = 'svg';

        const distDir = getDistPath(config.output, 'svg');
        const filename = getFilename(config.output, 'svg', isProd);
        const maxSize = config.output.dataUriLimit[assetType];

        const rule = chain.module.rule(CHAIN_ID.RULE.SVG).test(SVG_REGEX);

        // treat all .svg files as assets.
        chainStaticAssetRule({
          rule,
          maxSize,
          filename: path.posix.join(distDir, filename),
          assetType,
        });
      });
    },
  };
};
