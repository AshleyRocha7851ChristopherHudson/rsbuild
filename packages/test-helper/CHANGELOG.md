# @rsbuild/test-helper

## 0.2.0

### Patch Changes

- Updated dependencies [c11aa843]
- Updated dependencies [42622b45]
- Updated dependencies [4aa56abd]
- Updated dependencies [04deba66]
  - @rsbuild/core@0.2.0
  - @rsbuild/shared@0.2.0
