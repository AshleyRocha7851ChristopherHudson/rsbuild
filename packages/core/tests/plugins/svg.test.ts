import { createStubRsbuild } from '@rsbuild/test-helper';
import { pluginSvg } from '@src/plugins/svg';

describe('plugin-svg', () => {
  const cases = [
    {
      name: 'should allow using output.dataUriLimit.svg to custom data uri limit',
      rsbuildConfig: {
        output: {
          dataUriLimit: {
            svg: 666,
          },
        },
      },
    },
    {
      name: 'should allow to use distPath.svg to modify dist path',
      rsbuildConfig: {
        output: {
          distPath: {
            svg: 'foo',
          },
        },
      },
    },
    {
      name: 'should allow to use filename.svg to modify filename',
      rsbuildConfig: {
        output: {
          filename: {
            svg: 'foo.svg',
          },
        },
      },
    },
  ];

  it.each(cases)('$name', async (item) => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginSvg()],
      rsbuildConfig: item.rsbuildConfig,
    });

    const config = await rsbuild.unwrapConfig();

    expect(config).toMatchSnapshot();
  });
});
