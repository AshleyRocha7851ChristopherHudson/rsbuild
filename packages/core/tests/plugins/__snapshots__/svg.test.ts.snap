// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-svg > 'should allow to use distPath.svg to modify dist path' 1`] = `
{
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "generator": {
              "filename": "foo/[name].[contenthash:8].svg",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "foo/[name].[contenthash:8].svg",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 10000,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.svg\\$/,
      },
    ],
  },
}
`;

exports[`plugin-svg > 'should allow to use filename.svg to modify filename' 1`] = `
{
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/svg/foo.svg",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/svg/foo.svg",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 10000,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.svg\\$/,
      },
    ],
  },
}
`;

exports[`plugin-svg > 'should allow using output.dataUriLimit.svg to custom data uri limit' 1`] = `
{
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/svg/[name].[contenthash:8].svg",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/svg/[name].[contenthash:8].svg",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 666,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.svg\\$/,
      },
    ],
  },
}
`;
